### DISCUSSION TITLE
Feature Toggling

### DISCUSSION THREAD

**Discussion by lounah** [MobileNativeFoundation/discussions#77] [General]
Hi there!

How do you guys use feature-flags in your applications?
As of @TinkoffCreditSystems, toggles are a great instrument, which helps us in development and production.

Here's how:
There are two types of feature-toggles: **Local** and **Remote**.

The _local_ ones close the functionality only on debug builds, and do not add any impact to the release. This allows us to keep even incomplete features and whatever we want to keep in the main branch.
_Remote_ flags work like everywhere else - they just close the functionality for which we want to do some A/B testing stuff. We use a self-written service for remote flags managment.

So, when the development of any feature is started, this feature gets closed by a _Local_ feature-flag. So, we can merge feature-related pr's in a main branch and be insured, that it will not be in production. By the end of the development, we change the toggle type to a _Remote_, or remove this toggle from code.

The process of flags management is completely automated: we check the consistency of all toggles in our apps by using a bunch of self-written Gradle/Python scripts, which run on CI during regress.

How do you work with feature flags? 😄 
How do you manage the case, when the flag has been rolled out to 100% of users and needs to be removed from the code?

**Comment by waliahimanshu** [comment]
Hi, we have a similar approach of having local and remote toggles.
Local toggle helps us merge incomplete feature code in the main branch incrementally. And remote toggle from a webserver to do AB testing and turning the feature on and off.

Each feature toggle has some attributes like name, the owner (@slackHandle) and expiry date. 
We have a task that runs weekly, finds all expired toggles and posts a message on our slack channel notifying the owners. The owners then either extend the feature expiry date or removes it. For some permanent or long-running feature, we set the date too far in the future (20 30 years ahead)
